### PR TITLE
Add the sample() convenience function in CEL expression builder

### LINKED ISSUE #34805: New CEL convenience function: random()
*Title*: *New CEL convenience function: random()*

*Description*:
* When called, use `Random::RandomGenerator` to generate a random uint64
* Useful in [Access Logger Filter](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/access_loggers/filters/cel/v3/cel.proto#extensions-access-loggers-filters-cel-v3-expressionfilter) to introduce sampling
* Although this is possible with the Runtime filter, this would be useful for end users

*Questions*:
* Should this convenience function be available everywhere? E.g. [the access log formatter](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/formatter/cel/v3/cel.proto#extensions-formatter-cel-v3-cel)?
* On the flip note, should this function, and others in the future, be explicitly requested by the user? E.g. a list of extensions provided with a cel expression under a new type:
```
message CelActivationSettings {
  repeated google.protobuf.Any typed_functions = 1;
}

message CelRandomFunction {
}

...
```

### PR DESCRIPTION
**Pull request by wtzhang23**
Commit Message: add random convenience function in CEL expression builder
Additional Description: Add the `random()` function that allows CEL expressions to call the `Random::RandomGenerator::random()` function to generate a random 64-bit unsigned integer
Risk Level: Medium
Testing: unit tests
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a
Fixes #34805 
Fixes istio/istio#50404

### CODE REVIEW

**Comment by wtzhang23** [comment]
Might be more convenient to pass the Bernoulli value as a float instead of providing the uint64 value.

**Comment by wtzhang23** [comment]
Hmm, there's a bug where if constant folding is enabled, it will inline the function. Thinking of mitigations. Some things that can possibly be done:
* Close this PR and simply provide the random number through an attribute
* Change the function to be lazily evaluated (tried this initially)
* Push a change upstream to allow marking functions to not be inlined
Also, not sure if this function really aligns well with the cel-spec philosophy of functions being "side-effect free"

**Comment by tyxia** [comment]
> Hmm, there was a bug where if constant folding is enabled, it will inline the function. Thinking of mitigations. Some things that can possibly be done:
> 
> * Close this PR and simply provide the random number through an attribute
> * Change the function to be lazily evaluated (tried this initially but passing the Random::RandomGenerator at multiple locations became cumbersome. See my other [branch](https://github.com/envoyproxy/envoy/compare/main...wtzhang23:envoy:cel-random-lazy))
> * Push a change upstream to allow marking functions to not be inlined
>   Also, not sure if this function really aligns well with the cel-spec philosophy of functions being "side-effect free"

I don't think we should just disable constant folding (which is performance optimization) just for random feature. If it conflicts with your random feature, then this is first thing of this PR.
 + @kyessenov , Kuat's insights would be appreciated here.

**Comment by wtzhang23** [comment]
FWIW I do have another branch with lazy loading of functions that doesn't require disable constant folding.

**Comment by kyessenov** [comment]
I think either option (2) or (3) are reasonable. A new attribute only solves a subset of cases when you want the same consistent random value for all evaluations for a request.

@jcking might have a wider context on how randomness is added to CEL.

**Comment by wtzhang23** [comment]
It may still be useful to have (1), especially if it is based off of `x-request-id` in which case you can make a decision consistently across multiple envoy hops. Though, this is not related to this PR or the parent issue.

**Comment by kyessenov** [comment]
I'm looking into CEL solutions in other contexts - but I tend to agree that a Bernoulli value `sample(p)` that returns 1 with probability `p` might be easier to use. That solves some constant-folding issues but not entirely absolves non-deterministic evaluation. We can also make it integer-based using https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/v3/percent.proto#envoy-v3-api-msg-type-v3-fractionalpercent, e.g. `sample(p, 10000)` where p is between 0 and 10000.

**Comment by kyessenov** [comment]
I had an internal discussion with some CEL folks. We think a pure  `random()` function is likely to cause issues in general that won't be solved by lazy functions. CEL expressions generally are pre-processed in multiple places, not just Envoy, and a compiler at any place in the pipeline is free to perform transformations to improve performance. CEL functions are pure and deterministic, so if you have multiple calls to `random()` they can be elided and/or combined in unexpected ways. So even if you make evaluation assume lazy function in Envoy, the control plane can still make a dangerous transformation.

What we think would work best is to turn this into a true pure function by adding parameters. One example would be `sample(r, p, i)` where:
- `r` is an opaque attribute mapping to the random generator. User is expected to pass an attribute that is not known ahead of time, e.g. `request`.
- `p` is a Bernoulli ratio. For simplicity we can make this a numerator over 10000.
- `i` is a sampling count. This makes repeated sampling, e.g. in a `for` loop, explicit. If we only sample once per expression, we can elide this parameter.

**Comment by wtzhang23** [comment]
Ya, I was thinking a similar thing. To not be too computationally expensive, I was thinking under the hood, for each CEL expression, we generate one random number for each invocation of an expression (can be stored in the `r` attribute) that we use to seed some polynomial where we pass i.

**Comment by wtzhang23** [comment]
Thinking of an implementation where we generate a random number as a seed before the execution of a CEL expression, apply a cryptographic hash with that seed concatenated with a index (e.g. sha256), then convert the hash into a probability. So
`sample(r, p, i) = (sha256(r.seed + i)[..2] / u64max) < p`

**Comment by zirain** [comment]
As a user's point of view, I think this is a good proposal to implement percentage-based log sampling.

**Comment by wtzhang23** [comment]
Probably some optimization should be considered? E.g.
- [ ] Lazily generating a random number to serve as a seed to avoid generating a seed for every cel expression, even those that don't need it?
- [ ] Keep a small buffer for a small range of indices to avoid using the SHA technique when not necessary?

File: envoy/matcher/matcher.h [hunk 1]
```diff
@@ -285,7 +286,8 @@ template <class DataType> class DataInputFactory : public Config::TypedFactory {
    */
   virtual DataInputFactoryCb<DataType>
   createDataInputFactoryCb(const Protobuf::Message& config,
-                           ProtobufMessage::ValidationVisitor& validation_visitor) PURE;
+                           ProtobufMessage::ValidationVisitor& validation_visitor,
+                           Random::RandomGenerator& random) PURE;
```
Comment (wtzhang23) [inline envoy/matcher/matcher.h:290]: Probably should just pass the context?

**Comment by adisuissa** [comment]
Please merge main to fix conflicts.
/wait

**Comment by phlax** [comment]
@wtzhang23 coverage fail looks real

**Comment by wtzhang23** [comment]
Yep, need to get the coverage report locally to check. Unless it is already uploaded and I couldn't find the artifact.

**Comment by wtzhang23** [comment]
Bump on this? I no longer need this feature and I'd like to get this off my hands. I'm ok with closing this if it is not a good approach (I'm slightly uncomfortable with it to be honest, though I don't know of another way to conform to the CEL spec).

**Comment by kyessenov** [comment]
Sorry about the delay. I'm not confident this is the right approach TBH. I would be more comfortable if this was an extension, but we haven't added an ability to extend CEL vocabulary. I am worried that adding this without a concrete use case would hurt us long term, especially since you added functions to the top-level `request` attribute. But I also don't want this work go to waste. Should we first work out a custom vocabulary in CEL and then provide this as a custom extensions instead?

**Comment by wtzhang23** [comment]
Yes, I agree with your assessment.

**Comment by kyessenov** [comment]
/wait

Need some way to make this opt-in.

**Comment by martijneken** [comment]
Question about the approach taken here: why did this focus on random number generation, rather than hashing/fingerprinting some part of a request (e.g. one or more headers). The fingerprinting approach would allow users to key on a well known request attribute (e.g. a session ID or user ID), which avoids inlining concerns and leads to deterministic behavior across requests.

**Comment by wtzhang23** [comment]
Forgive me if I misspeak as my math is not the greatest, but I was concerned about if someone would like to sample independently. Though even in this approach, I'm not confident if subsequent calls are actually random. 

I agree that for the use case I was interested in (access log sampling), a single random value would be ok. A helper function converting `x-request-id` into a integer could work, and is what people seem to use today (e.g. check if the last character of the `x-request-id` is `A` and you get 1/16 sampling rate).

**Comment by martijneken** [comment]
Thanks for the helpful context! I understand what you were going for, and I'm glad to hear the hashing approach makes sense to you also.

**Pull request closed without merging**